src/tactic/rewrite_search/frontend.lean

```diff
@@ -0,0 +1,100 @@
+expressions that are equivalent to the left hand side, and one component contains expressions
+that are equivalent to the right hand side. The algorithm completes when we discover an
+edge that connects the two components, creating a path of rewrites that connects the
+left hand side and right hand side of the graph. For more detail, see Keeley's report at
+https://hoek.io/res/2018.s2.lean.report.pdf, although note that the edit distance mechanism
+described is currently not implemented, only plain breadth-first search.
+
+This algorithm is generally superior to one that only expands nodes starting from a single
+side, because it is replacing one tree of depth `2d` with two trees of depth `d`. This is
+a quadratic speedup for regular trees; our trees aren't regular but it's still probably
+a much better algorithm. We can only use this specific algorithm for rewrite-type tactics,
+though, not general sequences of tactics, because it relies on the fact that any rewrite
+can be reversed.
+
+## File structure
+
+* `discovery.lean` contains the logic for figuring out which rewrite rules to consider.
+* `search.lean` contains the graph algorithms to find a successful sequence of tactics.
+* `explain.lean` generates concise Lean code to run a tactic, from the autogenerated sequence
+  of tactics.
+* `frontend.lean` contains the user-facing interface to the `rewrite_search` tactics.
+* `types.lean` contains data structures shared across multiple of these components.
+-/
+
+namespace tactic.interactive
+
+open lean.parser interactive interactive.types tactic.rewrite_search
+
+/--
+Collects rewrite rules, runs a graph search to find a chain of rewrites to prove the
+current target, and generates a string explanation for it.
+-/
+private meta def rewrite_search_target (cfg : config) (rws : list rw_rule) :
+  tactic string :=
+do t ← tactic.target,
+  if t.has_meta_var then
+    tactic.fail "rewrite_search is not suitable for goals containing metavariables"
+  else tactic.skip,
+  rules ← collect_rules rws,
+  g ← mk_graph cfg rules t,
+  (_, proof, steps) ← g.find_proof,
+  tactic.exact proof >> (explain_search_result cfg rules proof steps)
+
+/-- A tactic to pick the default config. -/
+meta def pick_default : tactic unit := `[exact tactic.rewrite_search.default_config]
+
+/-- Search for a chain of rewrites to prove an equation or iff statement. -/
+meta def rewrite_search (try_harder : parse $ optional (tk "!"))
+  (cfg : config . pick_default) : tactic string :=
+rewrite_search_target cfg []
+
+add_tactic_doc
+{ name        := "rewrite_search",
+  category    := doc_category.tactic,
+  decl_names  := [`tactic.interactive.rewrite_search],
+  tags        := ["rewrite", "automation"] }
+
+/--
+Search for a chain of rewrites to prove an equation or iff statement.
+Includes the rewrite rules specified in the same way as the `rw` tactic accepts.
+-/
+meta def rewrite_search_with (try_harder : parse $ optional (tk "!")) (rs : parse rw_rules)
```

Comment:
Wouldn't it be nicer to match the syntax of `simp`, by having `rs` be an optional argument? So `by { rewrite_search }` works the same as it does now, and `by { rewrite_search [h1, h2, h3] }` is what is now `by { rewrite_search_with [h1, h2, h3] }`.

Reply:
That's a good idea, I'll take a stab at it.

Reply:
OK, this is in now. The code is a decent amount simpler. Many thanks to Rob's advice in Zulip ;-)